lib/semantics/check-do-concurrent.cc

```diff
@@ -0,0 +1,205 @@
+  }
+  // C1141
+  void Post(const parser::ProcedureDesignator &procedureDesignator) {
+    if (auto *name = std::get_if<parser::Name>(&procedureDesignator.u)) {
+      auto upperName{parser::ToUpperCaseLetters(name->ToString())};
+      if (upperName == "IEEE_GET_FLAG"s) {
+        messages_.Say(charBlock_,
+            parser::MessageFormattedText{
+                "IEEE_GET_FLAG not allowed in DO CONCURRENT"_err_en_US});
+      } else if (upperName == "IEEE_SET_HALTING_MODE"s) {
+        messages_.Say(charBlock_,
+            parser::MessageFormattedText{"IEEE_SET_HALTING_MODE not allowed"
+                                         " in DO CONCURRENT"_err_en_US});
+      } else if (upperName == "IEEE_GET_HALTING_MODE"s) {
+        messages_.Say(charBlock_,
+            parser::MessageFormattedText{"IEEE_GET_HALTING_MODE not allowed"
+                                         " in DO CONCURRENT"_err_en_US});
+      }
+    }
+  }
+
+  // C1138: extended ranges in DOs should be errors, not warnings
+
+  // 11.1.7.5
+  void Post(const parser::IoControlSpec &ioControlSpec) {
+    if (auto *charExpr =
+            std::get_if<parser::IoControlSpec::CharExpr>(&ioControlSpec.u)) {
+      if (std::get<parser::IoControlSpec::CharExpr::Kind>(charExpr->t) ==
+          parser::IoControlSpec::CharExpr::Kind::Advance) {
+        messages_.Say(charBlock_,
+            parser::MessageFormattedText{
+                "ADVANCE specifier not allowed in DO CONCURRENT"_err_en_US});
+      }
+    }
+  }
+
+private:
+  bool ObjectIsCoarray() { return false; } // placeholder
+  bool EndTDeallocatesCoarray() { return false; } // placeholder
+
+  parser::CharBlock charBlock_;
+  parser::Messages &messages_;
+};
+
+class DoConcurrentCollection {
+public:
+  DoConcurrentCollection(parser::Messages &messages) : messages_{messages} {}
+  template<typename T> constexpr bool Pre(const T &) { return true; }
+  template<typename T> constexpr void Post(const T &) {}
+  bool Pre(const parser::ExecutionPart &executionPart) {
+    const auto &cend{executionPart.v.cend()};
+    for (auto iter{executionPart.v.cbegin()}; iter != cend; ++iter) {
+      CheckDoConcurrent(iter);
+      currentIter_ = iter;
+    }
+    return true;
+  }
+  template<typename T> void Post(const parser::Statement<T> &statement) {
+    if (!labels_.empty() && statement.label.has_value() &&
+        labels_.back() == *statement.label) {
+      CheckConstraints(++labelDoIters_.back(), currentIter_);
+      labels_.pop_back();
+      labelDoIters_.pop_back();
+    }
+  }
+
+private:
+  void CheckConstraints(const parser::Block::const_iterator &begin,
+      const parser::Block::const_iterator &end) {
+    DoConcurrentEnforcement doConcurrentEnforcement{messages_};
+    for (auto iter = begin; iter != end; ++iter) {
+      Walk(*iter, doConcurrentEnforcement);
+    }
+  }
+  void CheckDoConcurrent(
+      const std::list<parser::ExecutionPartConstruct>::const_iterator &iter) {
```

Comment:
You can use the typedef `parser::Block` to save some typing.

Reply:
ok